### PR TITLE
MONGOID-5221/5222: Fix mongoization + demongoization of non-empty string as Integer/Float

### PR DESCRIPTION
**Pull request by johnnyshields**
This is a **partial** fix for [MONGOID-5222](https://jira.mongodb.org/browse/MONGOID-5222) and [MONGOID-5221](https://jira.mongodb.org/browse/MONGOID-5221)

### CODE REVIEW

File: lib/mongoid/criteria/queryable/extensions/numeric.rb [hunk 1]
```diff
@@ -51,7 +51,7 @@ module ClassMethods
             #
             # @return [ Object ] The converted number.
             def __numeric__(object)
-              object.to_s =~ /(\A[-+]?[0-9]+\z)|(\.0+\z)|(\.\z)/ ? object.to_i : Float(object)
+              object.to_s.match?(/\A[-+]?[0-9]*\.?0*\z/) ? object.to_i : Float(object)
```
Comment (johnnyshields) [inline lib/mongoid/criteria/queryable/extensions/numeric.rb:54]: Regexes are functionally equivalent. My new one is ~50% faster.

```ruby
Benchmark.bm do |bm|
  bm.report { x.times { '21329.2328'.match?(/(\A[-+]?[0-9]+\z)|(\.0+\z)|(\.\z)/) } }
  bm.report { x.times { '21329.2328'.match?(/\A[-+]?[0-9]*\.?0*\z/) } }
end
```

Comment (p-mongo) [reply lib/mongoid/criteria/queryable/extensions/numeric.rb:54]: They are not equivalent:

```
irb(main):001:0> /(\A[-+]?[0-9]+\z)|(\.0+\z)|(\.\z)/ =~ ''
=> nil
irb(main):002:0> /\A[-+]?[0-9]*\.?0*\z/ =~ ''
=> 0
```

Please make one change per PR.

Comment (johnnyshields) [reply lib/mongoid/criteria/queryable/extensions/numeric.rb:54]: I'll split this out.

Comment (johnnyshields) [reply lib/mongoid/criteria/queryable/extensions/numeric.rb]: Reverted

**Comment by johnnyshields** [comment]
@p-mongo I've clarified in the ticket the behavior should be identical for both mongoization and demongoization. There's no reason to convert the string "Foobar" to 0; it should be nil as it is non-sensical.

**Comment by johnnyshields** [comment]
@p-mongo please consider merging this at this time. While I appreciate there is an epic to fully review mongoization/demongoization, this PR is a "quick win" for a Float/Integer which are very commonly used.

**Comment by p-mongo** [comment]
I responded in https://jira.mongodb.org/browse/MONGOID-5222.

**Pull request closed without merging**